addons/mail/models/mail_message.py

```diff
@@ -1019,6 +1018,61 @@ def message_format(self, format_reply=True, msg_vals=None):
+        """ Prepare message to be personalized by partner.
+
+        This method add partner information in batch to the messages so that the
+        messages could be personalized for each partner by using
+        _message_format_personalize with no or a limited number of queries.
+
+        For example, it gathers all followers of the record related to the messages
+        in one go (or limit it to the partner given in parameter), so that the method
+        _message_format_personalize could then personalize each message for each partner.
+
+        :param list messages_formatted: list of message formatted using the method
+            message_format
+        :param list partner_ids: (optional) limit value computation to the partner
+            of the given list
+
+        :return: list of messages_formatted with added value:
+            'follower_id_by_partner_id': dict partner_id -> follower_id of the record
+        """
+        domain = expression.OR([
+            [('res_model', '=', model), ('res_id', 'in', list({value['res_id'] for value in values}))]
+            for model, values in
+            misc_groupby((vals for vals in messages_formatted if vals.get("model") and vals.get("res_id")),
+                         key=lambda r: r["model"])
+        ])
+        if partner_ids:
+            domain = expression.AND([domain, [('partner_id', 'in', partner_ids)]])
+        records_followed = self.env['mail.followers'].sudo().search(domain)
+        followers_by_record_ref = {(res_model, res_id): {value['partner_id'][0]: value['id'] for value in values}
+                                   for (res_model, res_id), values in
+                                   misc_groupby(records_followed.read(['res_id', 'res_model', 'partner_id']),
+                                                key=lambda r: (r["res_model"], r['res_id']))}
+        for vals in messages_formatted:
+            vals['follower_id_by_partner_id'] = followers_by_record_ref.get((vals['model'], vals['res_id']), dict())
+        return messages_formatted
+
+    def _message_format_personalize(self, partner_id, messages_formatted=None, format_reply=True, msg_vals=None):
+        """ Personalize the messages for the partner.
+
+        :param integer partner_id: id of the partner to personalize the messages for
+        :param list messages_formatted: (optional) list of message formatted using
+            the method _message_format_personalized_prepare.
+            If not provided message_format is called on self with provided parameters
+        :return: list of messages_formatted personalized for the partner
+        """
+        if not messages_formatted:
+            messages_formatted = self.message_format(format_reply=format_reply, msg_vals=msg_vals)
+            self._message_format_personalized_prepare(messages_formatted, [partner_id])
+        for vals in messages_formatted:
+            if 'follower_id_by_partner_id' in vals:
```

Comment:
Shouldn't we always set the value, even if it was not prepared (probably fallback to False) ? And in that case you can pop with a default value and remove the if, like
```
# set value for user being a follower, fallback to False if not prepared
follower_id_by_pid = vals.pop('follower_id_by_partner_id', {})
vals['user_follower_id'] = follower_id_by_pid.get(partner_id, False)
```